### PR TITLE
[#7444,#7465] Prevent replication with insufficient permissions (main)

### PR DESCRIPTION
**Pull request by alanking**
Addresses #7444 
Addresses #7465 

Need to run full test suite. New tests pass, of course.

### CODE REVIEW

File: scripts/irods/test/test_irepl.py [hunk 1]
```diff
@@ -1878,3 +1878,140 @@ def test_irepl_Sf_Rd_foo(self):
...
+                    self.assertTrue(lib.replica_exists_on_resource(self.user0, logical_path, self.target_resource))
+                    self.assertEqual(str(1), lib.get_replica_status(self.user0, os.path.basename(logical_path), 0))
+                    self.assertFalse(lib.replica_exists_on_resource(self.user0, logical_path, self.other_resource))
+
+                finally:
+                    # In the case where this test does not pass (i.e. REGRESSION) it is possible for the replicas to be
+                    # stuck in the intermediate or write-locked status. We set the status at the end here to ensure
+                    # that the object can be removed.
+                    with session.make_session_for_existing_admin() as admin_session:
+                        admin_session.assert_icommand(['ils', '-L', os.path.dirname(logical_path)], 'STDOUT')
```
Comment (trel) [inline scripts/irods/test/test_irepl.py:1937]: what does this assert?   same for those below that are identical.

Comment (alanking) [reply scripts/irods/test/test_irepl.py]: Just debuggin. I've added a comment for each.

**Review by trel — COMMENTED** [review]
smeagol!

**Comment by alanking** [comment]
Ran smack into https://github.com/irods/irods/issues/5992 with the GenQuery solution. I am able to get around this by kicking the query to a different spot where the `DATA_ID` is available. Alternatively, I could wait for GenQuery2 to become available and possibly use that.

Also realized that we will need tests for `irepl -a` since that is a different code path.

**Comment by alanking** [comment]
I also noticed that this is not going to work for remote users unless we include the `USER_ZONE` in the query. So, I need to add that, some federation tests, and `irepl -a` tests.

**Comment by alanking** [comment]
Marking ready for review again. Federation tests passed, running core tests now.

File: scripts/irods/lib.py [hunk 2]
```diff
@@ -684,14 +688,16 @@ def replica_exists(session, logical_path, replica_number):
...
+def replica_exists_on_resource(session, logical_path, resource_name, zone_name=None):
+    query = "select DATA_ID where COLL_NAME = '{}' and DATA_NAME = '{}' and DATA_RESC_NAME = '{}'".format(
+        os.path.dirname(logical_path),
+        os.path.basename(logical_path),
+        resource_name)
+    if zone_name:
+        command = ['iquest', '-z', zone_name, '%s', query]
+    else:
+        command = ['iquest', '%s', query]
+    return 'CAT_NO_ROWS_FOUND' not in session.run_icommand(command)[0]
```
Comment (trel) [inline scripts/irods/lib.py:703]: I found this line harder to read than I expected.   Perhaps a comment for the next reader?

Comment (alanking) [reply scripts/irods/lib.py:703]: Would it help to split up the line? Like this...
```suggestion
    out,_,_ = session.run_icommand(command)
    return 'CAT_NO_ROWS_FOUND' not in out
```

Comment (trel) [reply scripts/irods/lib.py:703]: i think it's the negative that got me... so either split them or a comment about what will happen for the True/False cases.

or... you know, assume the reader can read :P

Comment (trel) [reply scripts/irods/lib.py:703]: "that there no results" -> "that no results"

otherwise, this is very helpful - yes.  please.

**Comment by alanking** [comment]
Running core tests again now, for real.

**Comment by alanking** [comment]
Core tests passed

**Comment by alanking** [comment]
Unit and core tests passed

File: scripts/irods/test/test_federation.py [hunk 3]
```diff
@@ -2161,3 +2161,221 @@ def test_failure_to_local_from_remote__issue_6497(self):
 	def test_failure_to_local_from_local__issue_6497(self):
 		self.fail_to_overwrite_replica_on_target_resource_test_impl(copy_to='local', copy_from='local')
 
+
+class test_irepl_all_permission_levels__issue__7444_7465(unittest.TestCase):
```
Comment (korydraughn) [inline scripts/irods/test/test_federation.py:2165]: `__issue__` -> `__issue_`?

Wouldn't the extra underscore make grepping for `__issue_NNNN` more difficult?
Are there perks to including the extra underscore?

**Review by korydraughn — COMMENTED** [review]
Everything looks good. Just a few suggestions/questions.

File: scripts/irods/test/test_federation.py (hunk 3 above)
Comment (alanking) [reply scripts/irods/test/test_federation.py:2165]: That... is a typo, I think. Will remove the extra underscore.

File: server/api/src/rsDataObjRepl.cpp [hunk 4]
```diff
@@ -472,6 +477,62 @@ namespace
...
+                    [[fallthrough]];
+                case static_cast<int>(access_type::own):
+                    return true;
+
+                default:
+                    return false;
+            }
+        }
+        catch (const std::invalid_argument& e) {
+            log_api::error("DATA_ACCESS_TYPE for user [{}] on object [{}] has invalid value [{}]",
```
[the remark this answers is not in the document]
Comment (alanking) [reply server/api/src/rsDataObjRepl.cpp]: Done

File: server/api/src/rsDataObjRepl.cpp [hunk 5]
```diff
@@ -472,6 +477,62 @@ namespace
...
+                           _comm.clientUser.userName,
+                           _obj->logical_path(),
+                           access_type_string);
+            return false;
+        }
+        catch (const std::out_of_range& e) {
+            log_api::error("DATA_ACCESS_TYPE for user [{}] on object [{}] has out-of-range value [{}]",
```
[the remark this answers is not in the document]
Comment (alanking) [reply server/api/src/rsDataObjRepl.cpp]: Done

File: server/api/src/rsDataObjRepl.cpp [hunk 6]
```diff
@@ -692,6 +753,23 @@ namespace
         auto source_cond_input = irods::experimental::make_key_value_proxy(source_inp.condInput);
         auto source_obj = resolve_hierarchy_and_get_data_object_info(_comm, source_inp, irods::OPEN_OPERATION);
 
+        // This check is required because certain operations used by this API in order to perform a replication have
+        // different permissions requirements which may not align. In order to complete the replication, the maximum
+        // required permission level across all of these operations is the minimum required permission to perform the
+        // full operation. As such, an additional check is required to make sure the replication can proceed.
+        if (!user_has_permission_to_replicate_object(_comm, source_obj)) {
+            const std::string msg =
+                fmt::format("User [{}] does not have sufficient permission to replicate object [{}].",
```
[the remark this answers is not in the document]
Comment (alanking) [reply server/api/src/rsDataObjRepl.cpp]: Done

File: server/api/src/rsDataObjRepl.cpp [hunk 7]
```diff
@@ -862,6 +940,24 @@ namespace
...
         auto source_cond_input = irods::experimental::make_key_value_proxy(source_inp.condInput);
         auto source_obj = resolve_hierarchy_and_get_data_object_info(_comm, source_inp, irods::OPEN_OPERATION);
+
+        // This check is required because certain operations used by this API in order to perform a replication have
+        // different permissions requirements which may not align. In order to complete the replication, the maximum
+        // required permission level across all of these operations is the minimum required permission to perform the
+        // full operation. As such, an additional check is required to make sure the replication can proceed.
+        if (!user_has_permission_to_replicate_object(_comm, source_obj)) {
+            const std::string msg =
+                fmt::format("User [{}] does not have sufficient permission to replicate object [{}].",
```
[the remark this answers is not in the document]
Comment (alanking) [reply server/api/src/rsDataObjRepl.cpp]: Done

File: scripts/irods/test/test_federation.py (hunk 3 above)
Comment (alanking) [reply scripts/irods/test/test_federation.py]: Done

**Review by korydraughn — APPROVED** [review]
Squash/Pound it.

**Pull request merged**